pkg/abcupdater/metrics.go

```diff
@@ -0,0 +1,226 @@
+
+	"github.com/abcxyz/abc-updater/pkg/abcupdater/localstore"
+	"github.com/abcxyz/pkg/logging"
+)
+
+const (
+	installIDFileName = "id.json"
+)
+
+var regExUUID = regexp.MustCompile("^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$")
+
+type metricsConfig struct {
+	ServerURL string `env:"ABC_UPDATER_METRICS_URL,default=https://abc-updater-metrics.tycho.joonix.net"`
+}
+
+// MetricsInfo are the parameters for sending metrics.
+type MetricsInfo struct {
+	// The ID of the application to check.
+	AppID string
+
+	// The version of the app to check for updates.
+	// Should be of form vMAJOR[.MINOR[.PATCH[-PRERELEASE][+BUILD]]] (e.g., v1.0.1)
+	Version string
+
+	// An optional Lookuper to load envconfig structs. Will default to os environment variables.
+	Lookuper envconfig.Lookuper
+
+	Metrics map[string]int
+
+	// Optional override for install id file location. Mostly intended for testing.
+	// If empty uses default location.
+	InstallIDFileOverride string
+}
+
+// InstallIDData defines the json file that defines installation id.
+type InstallIDData struct {
+	// Time ID was created, in UTC epoch seconds.
+	IDCreatedTimestamp int64 `json:"idCreatedTimestamp"`
+	// InstallID. Expected to be a hex 8-4-4-4-12 formatted v4 UUID.
+	InstallID string `json:"installId"`
+}
+
+type SendMetricRequest struct {
+	// The ID of the application to check.
+	AppID string `json:"appId"`
+
+	// The version of the app to check for updates.
+	// Should be of form vMAJOR[.MINOR[.PATCH[-PRERELEASE][+BUILD]]] (e.g., v1.0.1)
+	Version string `json:"version"`
+
+	Metrics map[string]int `json:"metrics"`
+
+	// InstallID. Expected to be a hex 8-4-4-4-12 formatted v4 UUID.
+	InstallID string `json:"installId"`
+}
+
+// Stricter than uuid.Parse() which isn't meant for validating strings according
+// to documentation.
+func validInstallID(id string) bool {
+	return regExUUID.MatchString(id)
+}
+
+// SendMetricsSync sends information about application usage. Users can opt out
+// by setting an env variable defined in opt_out.go.
+// Accepts a context for cancellation.
+func SendMetricsSync(ctx context.Context, info *MetricsInfo) error {
+	lookuper := info.Lookuper
+	if lookuper == nil {
+		lookuper = envconfig.OsLookuper()
+	}
+
+	optOutSettings, err := loadOptOutSettings(ctx, lookuper, info.AppID)
+	if err != nil {
+		return fmt.Errorf("failed to load opt out settings: %w", err)
+	}
+
+	if optOutSettings.NoMetrics {
+		return nil
+	}
+
+	generateNewID := true
+	storedID, err := loadInstallID(info)
+	if err == nil && storedID != nil {
+		oneDayAgo := time.Now().Add(-24 * time.Hour)
+		// Defensively check for case ID Created is in future.
+		generateNewID = oneDayAgo.Unix() >= storedID.IDCreatedTimestamp ||
+			time.Now().Unix() < storedID.IDCreatedTimestamp
+	}
+	var installID string
+	if generateNewID {
+		installUUID, err := uuid.NewRandom()
+		if err != nil {
+			return fmt.Errorf("could not generate id for metrics: %w", err)
+		}
```

Comment:
```go
func randomID(l int) (string, error) {
  b := make([]byte, l)
  n, err := rand.Read(b)
  if err != nil {
    return "", fmt.Errorf("failed to read random bytes: %w", err)
  }
  if got, want := n, l; got != want {
    return "", fmt.Errorf("invalid random length, expected %d to be %d", got, want)
  }
  return hex.Encode(b), nil // or base64.StdEncoding.EncodeToString
}
```

Reply:
Took this, though checking `n` isn't necessary. According to docs:
```
On return, n == len(b) if and only if err == nil.
```

Reply:
Though actually there may be a bug in the docs or implementation:

https://cs.opensource.google/go/go/+/refs/tags/go1.22.3:src/crypto/rand/rand_unix.go

There is no shortcircuit from what I can tell, so wouldn't an error with a `b` of length 0 fail the `if and only if` statement?

Reply:
Meh - I've always checked to be safe.